velox/common/caching/SsdCache.cpp

```diff
@@ -0,0 +1,505 @@
+    evictLocked();
+  }
+}
+
+namespace {
+void addEntryToIovecs(AsyncDataCacheEntry& entry, std::vector<iovec>& iovecs) {
+  if (entry.tinyData()) {
+    iovecs.push_back({entry.tinyData(), static_cast<size_t>(entry.size())});
+    return;
+  }
+  auto& data = entry.data();
+  iovecs.reserve(iovecs.size() + data.numRuns());
+  int64_t bytesLeft = entry.size();
+  for (auto i = 0; i < data.numRuns(); ++i) {
+    auto run = data.runAt(i);
+    iovecs.push_back(
+        {run.data<char>(), std::min<size_t>(bytesLeft, run.numBytes())});
+    bytesLeft -= run.numBytes();
+    if (bytesLeft <= 0) {
+      break;
+    };
+  }
+}
+} // namespace
+
+SsdPin SsdFile::find(RawFileCacheKey key) {
+  SsdKey ssdKey{StringIdLease(fileIds(), key.fileNum), key.offset};
+  SsdRun run;
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    if (suspended_) {
+      return SsdPin();
+    }
+    newEventLocked();
+    auto it = entries_.find(ssdKey);
+    if (it == entries_.end()) {
+      return SsdPin();
+    }
+    run = it->second;
+    pinRegionLocked(run.offset());
+  }
+  return SsdPin(*this, run);
+}
+
+void SsdFile::newEventLocked() {
+  ++numEvents_;
+  if (numEvents_ > kDecayInterval && numEvents_ > entries_.size() / 2) {
+    numEvents_ = 0;
+    for (auto i = 0; i < numRegions_; ++i) {
+      int64_t score = regionScore_[i];
+      regionScore_[i] = (score * 15) / 16;
+    }
+  }
+}
+
+void SsdFile::load(SsdRun run, AsyncDataCacheEntry& entry) {
+  VELOX_CHECK_EQ(run.size(), entry.size());
+  regionScore_[regionIndex(run.offset())] += run.size();
+  ++stats_.entriesRead;
+  stats_.bytesRead += run.size();
+  if (entry.tinyData()) {
+    auto rc = pread(fd_, entry.tinyData(), run.size(), run.offset());
+    VELOX_CHECK_EQ(rc, run.size());
+  } else {
+    std::vector<struct iovec> iovecs;
+    addEntryToIovecs(entry, iovecs);
+
+    auto rc = folly::preadv(fd_, iovecs.data(), iovecs.size(), run.offset());
+    VELOX_CHECK_EQ(rc, run.size());
+  }
+  entry.setSsdFile(this, run.offset());
+}
+
+void SsdFile::read(
+    uint64_t offset,
+    const std::vector<folly::Range<char*>> buffers,
+    int32_t numEntries) {
+  stats_.entriesRead += numEntries;
+  uint64_t regionBytes = 0;
+  uint64_t regionOffset = offset;
+  for (auto range : buffers) {
+    if (range.data()) {
+      regionBytes += range.size();
+    }
+    regionOffset += range.size();
+  }
+  regionUsed(regionIndex(offset), regionBytes);
+  stats_.bytesRead += regionBytes;
+  readFile_->preadv(offset, buffers);
+}
+
+std::pair<uint64_t, int32_t> SsdFile::getSpace(
+    const std::vector<CachePin>& pins,
+    int32_t begin) {
+  std::lock_guard<std::mutex> l(mutex_);
+  for (;;) {
+    if (writableRegions_.empty()) {
+      if (!evictLocked()) {
+        return {0, 0};
+      }
+    }
+    auto region = writableRegions_[0];
+    auto offset = regionSize_[region];
+    auto available = kRegionSize - regionSize_[region];
+    int64_t toWrite = 0;
+    for (; begin < pins.size(); ++begin) {
```

Comment:
Is begin an index into the pins vector? Perhaps, clarify in the comments for this method in the header file.